repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/resource/LinkPreProcessor.java

```diff
@@ -1,55 +1,87 @@
 import org.apache.atlas.model.instance.EntityMutations;
+import org.apache.atlas.repository.graphdb.AtlasVertex;
+import org.apache.atlas.repository.store.graph.v2.EntityGraphRetriever;
 import org.apache.atlas.repository.store.graph.v2.EntityMutationContext;
+import org.apache.atlas.type.AtlasTypeRegistry;
 import org.apache.atlas.utils.AtlasPerfMetrics;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.util.regex.Pattern;
 
+import static org.apache.atlas.model.instance.EntityMutations.EntityOperation.CREATE;
+import static org.apache.atlas.model.instance.EntityMutations.EntityOperation.UPDATE;
+import static org.apache.atlas.repository.Constants.ASSET_LINK_EDGE_LABEL;
 import static org.apache.atlas.repository.Constants.ATTRIBUTE_LINK;
 import static org.apache.atlas.repository.Constants.QUALIFIED_NAME;
-public class LinkPreProcessor implements PreProcessor {
+
+public class LinkPreProcessor extends AbstractResourcePreProcessor {
     private static final Logger LOG = LoggerFactory.getLogger(LinkPreProcessor.class);
+
     private static final Pattern REGEX_ONSITE_URL = Pattern.compile("(?:[\\p{L}\\p{N}\\\\\\.\\#@\\$%\\+&;\\-_~,\\?=/!]+|\\#(\\w)+)");
     private static final Pattern REGEX_OFFSITE_URL = Pattern.compile("\\s*(?:(?:ht|f)tps?://|mailto:)[\\p{L}\\p{N}]"
             + "[\\p{L}\\p{N}\\p{Zs}\\.\\#@\\$%\\+&;:\\-_~,\\?=/!\\(\\)]*+\\s*");
     private static final Predicate<String> REGEX_ON_OFFSITE_URL = matchesEither(REGEX_ONSITE_URL, REGEX_OFFSITE_URL);
 
+    public LinkPreProcessor(AtlasTypeRegistry typeRegistry,
+                            EntityGraphRetriever entityRetriever) {
+        super(typeRegistry, entityRetriever);
+    }
+
     @Override
     public void processAttributes(AtlasStruct entityStruct, EntityMutationContext context, EntityMutations.EntityOperation operation) throws AtlasBaseException {
         if (LOG.isDebugEnabled()) {
             LOG.debug("LinkPreProcessor.processAttributes: pre processing {}, {}", entityStruct.getAttribute(QUALIFIED_NAME), operation);
         }
 
         AtlasEntity entity = (AtlasEntity) entityStruct;
-        String link = (String) entity.getAttribute(ATTRIBUTE_LINK);
 
         switch (operation) {
             case CREATE:
+                processLinkCreate(entity);
+                break;
             case UPDATE:
-                processLink(link, operation.name());
+                processLinkUpdate(entity, context.getVertex(entity.getGuid()));
                 break;
         }
 
     }
 
+    private void processLinkCreate(AtlasEntity linkEntity) throws AtlasBaseException {
+        validateLinkAttribute(linkEntity, CREATE.name());
+    }
+
+    private void processLinkUpdate(AtlasEntity linkEntity, AtlasVertex vertex) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("processLinkUpdate");
+
+        try {
+            validateLinkAttribute(linkEntity, UPDATE.name());
```

Comment:
Lets do authZ first

Reply:
I think it is better to validate the link value before checking for authz